### PR TITLE
Implement support for MongoDB Atlas fastsync

### PR DESCRIPTION
**Pull request by elephantum**
## Problem

`tap-mongodb` does not support MongoDB Atlas in fastsync mode.

## Proposed changes

Copy approach from `pipelinewise-tap-mongodb`: build `MongoClient` connection string both for API client and `mongodump`.

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [x] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [x] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

**Comment by elephantum** [comment]
See also: https://github.com/transferwise/pipelinewise-tap-mongodb/pull/37 for context

File: pipelinewise/fastsync/commons/tap_mongodb.py [hunk 1]
```diff
@@ -153,6 +154,54 @@ def transform_value(value: Any, path) -> Any:
...
+        connection_query['replicaSet'] = config['replica_set']
+
+    if use_ssl:
+        connection_query['tls'] = 'true'
+
+    # NB: "sslAllowInvalidCertificates" must ONLY be supplied if `SSL` is true.
+    if not verify_mode and use_ssl:
+        connection_query['tlsAllowInvalidCertificates'] = 'true'
+
+    # looks like mongo dump and pymongo differ in ssl params
```
Comment (Samira-El) [inline pipelinewise/fastsync/commons/tap_mongodb.py:186]: We can use `ssl`, mongodump expects `ssl` and pymongo supports both `tls` and `ssl` (`ssl` is alias for `tls`).

File: pipelinewise/fastsync/commons/tap_mongodb.py [hunk 2]
```diff
@@ -395,32 +428,21 @@ def _export_collection(self, export_dir: str, collection_name) -> str:
...
-                f'"{url}"',
-                '--forceTableScan',
-                '--gzip',
-                '-c',
-                collection_name,
-                '-o',
-                export_dir,
-            ]
-        )
+        LOGGER.info("Running: %s", cmd)
```
Comment (Samira-El) [inline pipelinewise/fastsync/commons/tap_mongodb.py:443]: Can you remove this line? we cannot log the command because it contains password in clear.

File: singer-connectors/tap-mongodb/requirements.txt [hunk 3]
```diff
@@ -1 +1,2 @@
-pipelinewise-tap-mongodb==1.2.0
+# We need unreleased support for "srv" property in order to sync with Atlas
```
Comment (Samira-El) [inline singer-connectors/tap-mongodb/requirements.txt:1]: Will release this new feature

File: Dockerfile [hunk 4]
```diff
@@ -1,26 +1,41 @@
...
-        mongo-tools \
         mbuffer \
         wget \
+        git \
     && rm -rf /var/lib/apt/lists/* \
-    && pip install -U --no-cache-dir pip \
+    && pip install -U --no-cache-dir pip
+
+# In order to use fastsync with MongoDB Atlas we need version 100+ of mongodump
+RUN cd /tmp && \
```
Comment (Samira-El) [inline Dockerfile:17]: Can you make same change to this entrypoint [script](https://github.com/transferwise/pipelinewise/blob/master/dev-project/entrypoint.sh), it's the one used for dev and testing purposes.

File: setup.py [hunk 5]
```diff
@@ -36,7 +36,8 @@
           'tzlocal>=2.0,<4.1',
           'slackclient>=2.7,<2.10',
           'psutil==5.8.0',
-          'ujson>=4.1,<4.3'
+          'ujson>=4.1,<4.3',
+          'dnspython>=2.0,<2.2',
```
Comment (Samira-El) [inline setup.py:40]: can you pin this to a specific minor version, eg `2.1.*`, it's more deterministic.

Comment (elephantum) [reply setup.py:40]: Should we do the same in `pipelinewise-tap-mongodb`? This requirement is a copy-paste from there: https://github.com/transferwise/pipelinewise-tap-mongodb/blob/master/setup.py#L21

Comment (Samira-El) [reply setup.py:40]: yep

File: Dockerfile [hunk 6]
```diff
@@ -1,26 +1,41 @@
...
 
 RUN apt-get -qq update \
     && apt-get -qqy --no-install-recommends install \
         apt-utils \
         alien \
         libaio1 \
-        mongo-tools \
         mbuffer \
         wget \
+        git \
```
Comment (Samira-El) [inline Dockerfile:12]: why is this needed?

Comment (elephantum) [reply Dockerfile:12]: It was needed to install `tap-mongodb` from `master` branch. It should go away as soon as you release updated version with `srv` support.

File: Dockerfile (hunk 4 above)
Comment (elephantum) [reply Dockerfile:17]: @Samira-El in dev env you use mongo PPA, I updated version and used package from PPA

**Comment by Samira-El** [comment]
Can you update https://github.com/transferwise/pipelinewise/blob/master/Dockerfile.barebone too with same changes as https://github.com/transferwise/pipelinewise/blob/master/Dockerfile

File: dev-project/entrypoint.sh [hunk 7]
```diff
@@ -14,7 +14,7 @@ apt-get install -y --no-install-recommends \
   libaio1 \
   mariadb-client \
   mbuffer \
-  mongo-tools \
+  mongodb-database-tools \
```
Comment (Samira-El) [inline dev-project/entrypoint.sh:17]: This package doesn't exist

Comment (elephantum) [reply dev-project/entrypoint.sh:17]: Let me check

File: dev-project/entrypoint.sh [hunk 8]
```diff
@@ -4,7 +4,7 @@ set -e
 
 # Add Mongodb ppa
 apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 4B7C549A058F8B6B
-echo "deb [ arch=amd64 ] https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 multiverse" | tee /etc/apt/sources.list.d/mongodb.list
+echo "deb [ arch=amd64 ] https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.4 multiverse" | tee /etc/apt/sources.list.d/mongodb.list
```
Comment (Samira-El) [inline dev-project/entrypoint.sh:7]: We can't change this, the gpg key above is linked to this particular version

**Comment by elephantum** [comment]
@Samira-El is there a reason for Dockerfile.barebone to do all the setup work in script instead of Dockerfile?

File: dev-project/entrypoint.sh (hunk 8 above)
Comment (elephantum) [reply dev-project/entrypoint.sh:7]: Probably that's the reason for missing package.

It would be cleaner to upgrade mongo version. If it's not possible, we can port changes from Dockerfile

**Comment by Samira-El** [comment]
> @Samira-El is there a reason for Dockerfile.barebone to do all the setup work in script instead of Dockerfile?

Dockerfile.barebone is used to build a barebone image of pipelinewise that doesn't have any connectors installed.
This repos builds and publishes two images.

File: dev-project/entrypoint.sh (hunk 8 above)
Comment (Samira-El) [reply dev-project/entrypoint.sh:7]: Let's upgrade to 4.4, can you replace line 6 with `wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | apt-key add -`

Comment (Samira-El) [reply dev-project/entrypoint.sh:7]: line 8 also needs update to remove `4.2.7`

Comment (elephantum) [reply dev-project/entrypoint.sh:7]: Actually, this is cleaner than Dockerfile, I think it would be reasonable to port it to the main Dockerfile as well

Comment (Samira-El) [reply dev-project/entrypoint.sh:7]: yeah

File: singer-connectors/tap-mongodb/requirements.txt (hunk 3 above)
Comment (Samira-El) [reply singer-connectors/tap-mongodb/requirements.txt:1]: @elephantum 1.3.0 is out:
https://pypi.org/project/pipelinewise-tap-mongodb/1.3.0/

Comment (Samira-El) [reply singer-connectors/tap-mongodb/requirements.txt:1]: Please try it out.

**Comment by elephantum** [comment]
@Samira-El I think everything looks fine. 1.3.0 works for me, I've updated PR to use this version

File: Dockerfile [hunk 9]
```diff
@@ -1,26 +1,43 @@
...
         libaio1 \
-        mongo-tools \
         mbuffer \
         wget \
     && rm -rf /var/lib/apt/lists/* \
-    && pip install -U --no-cache-dir pip \
+    && pip install -U --no-cache-dir pip
+
+# Add Mongodb ppa
+RUN wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | apt-key add - \
```
Comment (Samira-El) [inline Dockerfile:17]: Can you combine this `RUN` with the one above?

Comment (elephantum) [reply Dockerfile:17]: Yes, but it will not reduce a number of `apt-get update/install` commands because `slim` image does not have `wget` and `gnupg` pre-installed.

Comment (elephantum) [reply Dockerfile:17]: Given that it seemed to me cleaner to separate them into different cmds

Comment (Samira-El) [reply Dockerfile:17]: Ah right!

**Comment by Samira-El** [comment]
Looks fine, one last change, can you please update this doc https://github.com/transferwise/pipelinewise/blob/master/docs/connectors/taps/mongodb.rst  to mention that we have support for MongoDb Atlas and how to enable it?

**Comment by elephantum** [comment]
> Looks fine, one last change, can you please update this doc https://github.com/transferwise/pipelinewise/blob/master/docs/connectors/taps/mongodb.rst to mention that we have support for MongoDb Atlas and how to enable it?

I did my best

**Review by Samira-El — APPROVED** [review]
